craft_archives/repo/apt_preferences_manager.py

```diff
@@ -0,0 +1,140 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2015-2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""Manage the host's apt preferences configurations."""
+import dataclasses
+import io
+import logging
+import typing
+from pathlib import Path
+
+_DEFAULT_PREFERENCES_FILE = Path("/etc/apt/preferences.d/craft-archives")
+
+_DEFAULT_HEADER = "# This file is managed by craft-archives"
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class Preference:
+    """A single preference paragraph.
+
+    Matches preferences as defined by apt:
+    https://manpages.debian.org/bullseye/apt/apt_preferences.5.en.html
+    """
+
+    pin: str
+    priority: int
+
+    @classmethod
+    def from_string(cls, input_str: str) -> "Preference":
+        """Create a Preference object from a preferences compliant string.
+
+        :param input_str: The string containing a preferences paragraph
+        :returns: a single Preference object
+        :raises: ValueError if unknown
```

Comment:
maybe we should use a more specific error class? It doesn't need to be _too_ specific, but I think something that has `ArchivesError` as superclass is a good idea.

Reply:
Good idea! I've created an AptPreferencesError for it.